### ISSUE TITLE
invalid_client error returned while trying to sign in with web UI

### ISSUE THREAD

**Issue by marcinax** [aws-amplify/amplify-swift#2799]
### Describe the bug

When trying to sign using web UI in Swift:
`try await Amplify.Auth.signInWithWebUI(for: provider, presentationAnchor: window)`
I always get an `{"error":"invalid_client"}` from Amplify SDK. 

Seems like my amplify config json is just fine because this configuration currently works on Android and on iOS I'm able to open web UI, sign in with Apple or Google and go back to the app (web view closes). Then, an error mentioned above occurs on iOS. Email/Password login also works fine - "invalid_client" happens only with web UI.
I've spent some time investigating this issue and pinned a proxy to verify the requests. It seems that SDK does not add any Authorization header to the `POST oauth2/token` request so there is a 400 (Bad Request) in the response.

### Steps To Reproduce

```swift
Steps to reproduce the behavior:
1. Use `signInWithWebUI` and set provider to Apple or Google.
2. Web view opens.
3. Authenticate with selected provider in web view.
4. Web view closes.
5. SDK returns 400 invalid_client error.
```

### Expected behavior

I want to sign in successfully with web UI using Apple or Google providers. In other words, it's necessary to get a success response from `oauth2/token` endpoint triggered by the SDK.

### Amplify Framework Version

2.3.2

### Amplify Categories

Auth

### Dependency manager

Swift PM

### Swift version

5

### CLI version

aws-amplify-cli/2.0

### Xcode version

14.2

### Relevant log output

```shell
<details>
<summary>Log Messages</summary>


INSERT LOG MESSAGES HERE
```
</details>
```


### Is this a regression?

No

### Regression additional context

_No response_

### Device

iPhone 14

### iOS Version

iOS 16

### Specific to simulators

_No response_

### Additional context

_No response_

**Comment by ruisebas** [comment]
Hi @marcinax , thanks for opening this issue.

`invalid_client` usually means some sort of client authentication failure. In the past, we've seen it happening when there is a mismatch in what's defined in the configuration file and what's present in the Cognito User Pools console, for example:
1. An `"AppClientSecret"` key present in the configuration file with no client secret is defined for that app client in the console.
1. The `"Scopes"` listed under "OAuth" contains scopes are not enabled for that app client in the console.

Could you please validate that those values match and try again? 

If all looks good or that does not fix your issue, would you mind sharing your configuration file with all sensitive information removed? That could help us investigate further.

Thanks!

**Comment by marcinax** [comment]
Thank you for the answer @ruisebas,
We made sure that configuration is ok. To prove that I've switched back to [1.29.0](https://github.com/aws-amplify/amplify-swift/releases/tag/v1.29.0) (last version before v2) and sign in with Apple and Google worked as expected. 
Then, switched again to 2.4.1 and problem still occurs. Tried also with the newest (2.5.0) - still doesn't work as expected. Also, with 2.5.0 I can successfully login with email/password but the problem is with web UI methods.
Can you make sure that there is no issue/error in the SDK v2+?

**Comment by ruisebas** [comment]
Thanks for your answer, we will attempt to reproduce this issue locally and provide any updates in here.

In the meantime and to help us debug further, would you be able to share the verbose logs you get when this issue happens for you? 
You can enable verbose logging with 

```swift
Amplify.Logging.logLevel = .verbose
```

Be sure to remove any sensitive information before posting them.
Thanks!

**Comment by marcinax** [comment]
Attaching logs:
```
2023-03-10 08:10:26.119469+0100 AppName[47849:14227233] [AWSAuthWebUISignInTask] Starting execution
2023-03-10 08:10:26.119551+0100 AppName[47849:14227233] [AWSAuthTaskHelper] Check if authstate configured
2023-03-10 08:10:26.119705+0100 AppName[47849:14227233] [AWSAuthTaskHelper] Auth state configured
2023-03-10 08:10:26.122378+0100 AppName[47849:14227233] [HostedUISignInHelper] Wait for a valid state
2023-03-10 08:10:26.122463+0100 AppName[47849:14227233] [HostedUISignInHelper] Start signIn flow
2023-03-10 08:10:26.122811+0100 AppName[47849:14227233] [HostedUISignInHelper] Wait for signIn to complete
2023-03-10 08:10:26.122889+0100 AppName[47849:14227230] [AWSCognitoAuthPlugin] Auth state change:

{
    "AuthState.configured" =     {
        "AuthenticationState.signingIn" =         {
            "SignInState.notStarted" =             {
            };
        };
        "AuthorizationState.configured" =         {
        };
    };
}
2023-03-10 08:10:26.123350+0100 AppName[47849:14227391] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/IntializeSignInFlow.swift Starting execution
2023-03-10 08:10:26.123404+0100 AppName[47849:14227391] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/IntializeSignInFlow.swift Sending event SignInEvent.initiateHostedUISignIn
2023-03-10 08:10:26.127010+0100 AppName[47849:14227233] [UserPoolSignInHelper] Checking next step for: notStarted
2023-03-10 08:10:26.127190+0100 AppName[47849:14227230] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/InitializeHostedUISignIn.swift Starting execution
2023-03-10 08:10:26.127193+0100 AppName[47849:14227233] [AWSCognitoAuthPlugin] Auth state change:

{
    "AuthState.configured" =     {
        "AuthenticationState.signingIn" =         {
            "SignInState.signingInWithHostedUI" =             {
                "HostedUISignInState.notStarted" =                 {
                };
            };
        };
        "AuthorizationState.configured" =         {
        };
    };
}
2023-03-10 08:10:26.128200+0100 AppName[47849:14227391] [AWSCognitoAuthPlugin] Credential Store state change:

loadingStoredCredentials
2023-03-10 08:10:26.128258+0100 AppName[47849:14227230] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/LoadCredentialStore.swift Starting execution
2023-03-10 08:10:26.129430+0100 AppName[47849:14227225] [UserPoolSignInHelper] Checking next step for: signingInWithHostedUI(AWSCognitoAuthPlugin.HostedUISignInState.notStarted)
2023-03-10 08:10:26.130357+0100 AppName[47849:14227230] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/LoadCredentialStore.swift Retreiving credential deviceMetadata(username: "unknown")
2023-03-10 08:10:26.148758+0100 AppName[47849:14227230] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/LoadCredentialStore.swift Sending event CredentialStoreEvent.throwError
2023-03-10 08:10:26.148893+0100 AppName[47849:14227225] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/IdleCredentialStore.swift Starting execution
2023-03-10 08:10:26.148915+0100 AppName[47849:14227225] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/IdleCredentialStore.swift Sending event CredentialStoreEvent.moveToIdleState
2023-03-10 08:10:26.148948+0100 AppName[47849:14227233] [AWSCognitoAuthPlugin] Credential Store state change:

error(KeychainStoreError: Unable to find the keychain item
Recovery suggestion: This should not happen. There is a possibility that there is a bug if this error persists. Please take a look at https://github.com/aws-amplify/amplify-ios/issues to see if there are any existing issues that match your scenario, and file an issue with the details of the bug if there isn't. Issue encountered at:
file: /Users/username/Library/Developer/Xcode/DerivedData/AppName-alpbkvcwukblukfnyozemhopxmnh/SourcePackages/checkouts/amplify-ios/AmplifyPlugins/Core/AWSPluginsCore/Keychain/KeychainStoreError.swift
function: recoverySuggestion
line: 69)
2023-03-10 08:10:26.148994+0100 AppName[47849:14227233] [AWSCognitoAuthPlugin] Credential Store state change:

idle
2023-03-10 08:10:26.149011+0100 AppName[47849:14227230] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/LoadCredentialStore.swift Starting execution
2023-03-10 08:10:26.149016+0100 AppName[47849:14227233] [AWSCognitoAuthPlugin] Credential Store state change:

loadingStoredCredentials
2023-03-10 08:10:26.149092+0100 AppName[47849:14227230] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/LoadCredentialStore.swift Retreiving credential asfDeviceId(username: "unknown")
2023-03-10 08:10:26.153166+0100 AppName[47849:14227230] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/LoadCredentialStore.swift Sending event CredentialStoreEvent.completedOperation
2023-03-10 08:10:26.153255+0100 AppName[47849:14227230] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/IdleCredentialStore.swift Starting execution
2023-03-10 08:10:26.153271+0100 AppName[47849:14227230] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/IdleCredentialStore.swift Sending event CredentialStoreEvent.moveToIdleState
2023-03-10 08:10:26.155505+0100 AppName[47849:14227391] [AWSCognitoAuthPlugin] Credential Store state change:

success(AWSCognitoAuthPlugin.CredentialStoreData.asfDeviceId("{HIDDEN DATA}", "unknown"))
2023-03-10 08:10:26.155544+0100 AppName[47849:14227391] [AWSCognitoAuthPlugin] Credential Store state change:

idle
2023-03-10 08:10:26.156136+0100 AppName[47849:14227230] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/InitializeHostedUISignIn.swift Sending event HostedUIEvent.showHostedUI
2023-03-10 08:10:26.156298+0100 AppName[47849:14227230] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/ShowHostedUISignIn.swift Starting execution
2023-03-10 08:10:26.156335+0100 AppName[47849:14227230] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/ShowHostedUISignIn.swift Showing url https://AppName-dev.auth.us-east-1.amazoncognito.com/oauth2/authorize?response_type=code&code_challenge_method=S256&client_id={HIDDEN DATA}&state={HIDDEN DATA}&redirect_uri=AppName://&scope=aws.cognito.signin.user.admin%20email%20openid%20phone%20profile&code_challenge={HIDDEN DATA}&userContextData={HIDDEN DATA} 
08:10:26.156354+0100 AppName[47849:14227391] [AWSCognitoAuthPlugin] Auth state change:

{
    "AuthState.configured" =     {
        "AuthenticationState.signingIn" =         {
            "SignInState.signingInWithHostedUI" =             {
                "HostedUISignInState.showingUI" =                 {
                };
            };
        };
        "AuthorizationState.configured" =         {
        };
    };
}
{HIDDEN DATA}%3D&identity_provider=SignInWithApple
2023-03-10 08:10:26.160920+0100 AppName[47849:14227233] [UserPoolSignInHelper] Checking next step for: signingInWithHostedUI(AWSCognitoAuthPlugin.HostedUISignInState.showingUI(AWSCognitoAuthPlugin.HostedUISigningInState(signInURL: https://AppName-dev.auth.us-east-1.amazoncognito.com/oauth2/authorize?response_type=code&code_challenge_method=S256&client_id={HIDDEN DATA}&state={HIDDEN DATA}9&redirect_uri=AppName://&scope=aws.cognito.signin.user.admin%20email%20openid%20phone%20profile&code_challenge={HIDDEN DATA}&userContextData={HIDDEN DATA}%3D&identity_provider=SignInWithApple, state: "{HIDDEN DATA}", codeChallenge: "{HIDDEN DATA}", presentationAnchor: Optional(<UIWindow: 0x106e18680; frame = (0 0; 393 852); gestureRecognizers = <NSArray: 0x282a1b180>; layer = <UIWindowLayer: 0x282a1b000>>), options: AWSCognitoAuthPlugin.HostedUIOptions(scopes: ["aws.cognito.signin.user.admin", "email", "openid", "phone", "profile"], providerInfo: AWSCognitoAuthPlugin.HostedUIProviderInfo(authProvider: Optional(Amplify.AuthProvider.apple), idpIdentifier: nil), presentationAnchor: Optional(<UIWindow: 0x106e18680; frame = (0 0; 393 852); gestureRecognizers = <NSArray: 0x282a1b180>; layer = <UIWindowLayer: 0x282a1b000>>), preferPrivateSession: false))))
2023-03-10 08:10:50.162460+0100 AppName[47849:14227532] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/ShowHostedUISignIn.swift Sending event HostedUIEvent.fetchToken
2023-03-10 08:10:50.169081+0100 AppName[47849:14227532] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/FetchHostedUISignInToken.swift Starting execution
2023-03-10 08:10:50.169275+0100 AppName[47849:14227225] [AWSCognitoAuthPlugin] Auth state change:

{
    "AuthState.configured" =     {
        "AuthenticationState.signingIn" =         {
            "SignInState.signingInWithHostedUI" =             {
                "HostedUISignInState.fetchingToken" =                 {
                };
            };
        };
        "AuthorizationState.configured" =         {
        };
    };
}
2023-03-10 08:10:50.173542+0100 AppName[47849:14227391] [UserPoolSignInHelper] Checking next step for: signingInWithHostedUI(AWSCognitoAuthPlugin.HostedUISignInState.fetchingToken)
2023-03-10 08:10:50.857750+0100 AppName[47849:14227391] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/FetchHostedUISignInToken.swift Sending event HostedUIEvent.throwError
2023-03-10 08:10:50.858267+0100 AppName[47849:14227391] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/CancelSignIn.swift Starting execution
2023-03-10 08:10:50.866469+0100 AppName[47849:14227391] [AWSCognitoAuthPlugin] AWSCognitoAuthPlugin/CancelSignIn.swift Sending event AuthenticationEvent(id: "774CA84F-ABF9-4126-9112-BB11E3766472", eventType: AWSCognitoAuthPlugin.AuthenticationEvent.EventType.cancelSignIn, time: nil)
2023-03-10 08:10:50.867274+0100 AppName[47849:14227545] [UserPoolSignInHelper] Checking next step for: signingInWithHostedUI(AWSCognitoAuthPlugin.HostedUISignInState.error(AWSCognitoAuthPlugin.SignInError.hostedUI(AWSCognitoAuthPlugin.HostedUIError.serviceMessage("invalid_client "))))
2023-03-10 08:10:50.867391+0100 AppName[47849:14227532] [AWSCognitoAuthPlugin] Auth state change:

{
    "AuthState.configured" =     {
        "AuthenticationState.signingIn" =         {
            "SignInState.signingInWithHostedUI" =             {
                "HostedUISignInState.error" =                 {
                    error = "AWSCognitoAuthPlugin.SignInError.hostedUI(AWSCognitoAuthPlugin.HostedUIError.serviceMessage(\"invalid_client \"))";
                };
            };
        };
        "AuthorizationState.configured" =         {
        };
    };
}
2023-03-10 08:10:50.867584+0100 AppName[47849:14227532] [AWSCognitoAuthPlugin] Auth state change:

{
    "AuthState.configured" =     {
        "AuthenticationState.signedOut" =         {
            lastKnownUserName = "(nil)";
        };
        "AuthorizationState.configured" =         {
        };
    };
}
2023-03-10 08:10:50.869834+0100 AppName[47849:14227545] [HostedUISignInHelper] Sending cancel signIn
2023-03-10 08:10:50.869953+0100 AppName[47849:14227545] [HostedUISignInHelper] Wait for signIn to cancel

```

**Comment by aryakrishn** [comment]
I am also experiencing the same issue. Is there any solution to this problem?

**Comment by ruisebas** [comment]
@marcinax @aryakrishn thanks for the details.

We've just released Amplify [2.6.0](https://github.com/aws-amplify/amplify-swift/releases/tag/2.6.0), which includes a fix for this issue.  Please let us know if you're still facing the issue after updating your dependencies.

We apologize for the inconveniences this caused.
